src/signature.c

```diff
@@ -118,145 +118,132 @@ void terminate_signature(void)
 	CRYPTO_cleanup_all_ex_data();
 }
 
-/* Verifies that the file and the signature exists, and does a signature
- * check afterwards.
+/* Verifies that the file and the signature exists, and does a signature check
+ * afterwards. If any error is to be considered a verify failure, then
+ * ERRORS_FATAL should be set to true.
  *
  * returns: true if able to validate the signature, false otherwise */
-static bool verify_signature(const char *data_filename, const char *sig_filename)
+static bool verify_signature(const char *data_filename, const char *sig_filename, bool errors_fatal)
 {
 	int ret;
+	bool result = false;
 	struct stat st;
+	char *errorstr = NULL;
 
-	int data_fd;
+	int data_fd = -1;
 	size_t data_len;
 	unsigned char *data = NULL;
-	BIO *data_BIO;
+	BIO *data_BIO = NULL;
 
-	int sig_fd;
+	int sig_fd = -1;
 	size_t sig_len;
 	unsigned char *sig = NULL;
-	BIO *sig_BIO;
+	BIO *sig_BIO = NULL;
 
-	PKCS7 *p7;
-	BIO *verify_BIO;
+	PKCS7 *p7 = NULL;
+	BIO *verify_BIO = NULL;
 
 	/* get the signature */
 	sig_fd = open(sig_filename, O_RDONLY);
```

Comment:
Why open() before fstat()'ing the file? The opposite order would make more sense IMO